eventsourcing/src/main/java/org/axonframework/eventsourcing/AnnotationBasedEventStateApplier.java

```diff
@@ -0,0 +1,117 @@
+import org.axonframework.messaging.annotation.AnnotatedHandlerInspector;
+import org.axonframework.messaging.annotation.ClasspathHandlerDefinition;
+import org.axonframework.messaging.annotation.ClasspathParameterResolverFactory;
+import org.axonframework.messaging.unitofwork.ProcessingContext;
+
+import java.util.concurrent.CompletableFuture;
+import javax.annotation.Nonnull;
+
+import static java.util.Objects.requireNonNull;
+
+
+/**
+ * Implementation of {@link EventStateApplier} that applies state changes through {@link EventSourcingHandler} annotated
+ * methods using an {@link AnnotatedHandlerInspector}.
+ *
+ * @param <M> The type of model to apply state changes to
+ * @author Mateusz Nowak
+ * @see EventSourcingHandler
+ * @see AnnotatedHandlerInspector
+ * @since 5.0.0
+ */
+public class AnnotationBasedEventStateApplier<M> implements EventStateApplier<M> {
+
+    private final AnnotatedHandlerInspector<Object> inspector;
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param modelType The type of model this instance will handle state changes for.
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull Class<M> modelType) {
+        this(
+                modelType,
+                AnnotatedHandlerInspector.inspectType(modelType,
+                                                      ClasspathParameterResolverFactory.forClass(modelType),
+                                                      ClasspathHandlerDefinition.forClass(modelType))
+        );
+    }
+
+    /**
+     * Initialize a new {@link AnnotationBasedEventStateApplier}.
+     *
+     * @param modelType The type of model this instance will handle state changes for.
+     * @param inspector The inspector to use to find the annotated handlers on the model.
+     */
+    public AnnotationBasedEventStateApplier(@Nonnull Class<M> modelType,
+                                            @Nonnull AnnotatedHandlerInspector<Object> inspector
+    ) {
+        requireNonNull(modelType, "Model Type may not be null");
+        this.inspector = requireNonNull(inspector,
+                                        "Annotated Handler Inspector may not be null");
+    }
+
+    @Override
+    public M apply(@Nonnull M model, @Nonnull EventMessage<?> event, @Nonnull ProcessingContext processingContext) {
+        requireNonNull(model, "Model may not be null");
+        requireNonNull(event, "Event Message may not be null");
+
+        try {
+            var eventHandlerResult = handle(model, event, processingContext).join();
+            return newModelInstance(model, eventHandlerResult);
+        } catch (Exception e) {
+            throw new StateEvolvingException(
+                    "Failed to apply event [" + event.type() + "] in order to evolve [" + model.getClass() + "] state",
+                    e);
+        }
+    }
+
+    private M newModelInstance(M model, MessageStream.Entry<?> eventHandlerResult) {
```

Comment:
I was confused by what this method did.  I'd suggest a more clearly intentional name

```suggestion
    private M modelFromStreamResultOrExisting(M existing, MessageStream.Entry<?> potentialModelFromStream) {
```